netbsd/Platform.c

```diff
@@ -323,7 +330,100 @@ bool Platform_getNetworkIO(NetworkIOData* data) {
-   // TODO
-   (void)percent;
-   (void)isOnAC;
+   int fd;
+   bool isACAD, isBattery;
+   int64_t isPresent, isConnected;
+   int64_t curCharge, maxCharge;
+   int64_t totalCharge, totalCapacity;
+   prop_dictionary_t dict, fields, props;
+   prop_object_iterator_t devIter, fieldsIter;
+   prop_object_t device, class, fieldsArray, curValue, maxValue, descField;
+
+   totalCharge = 0;
+   totalCapacity = 0;
+
+   *percent = NAN;
+   *isOnAC = AC_ERROR;
+
+   fd = open(_PATH_SYSMON, O_RDONLY);
+   if (fd == -1)
+      goto error;
+
+   if (prop_dictionary_recv_ioctl(fd, ENVSYS_GETDICTIONARY, &dict) != 0)
+      goto error;
+
+   devIter = prop_dictionary_iterator(dict);
+   if (devIter == NULL)
+      goto error;
+
+   while ((device = prop_object_iterator_next(devIter)) != NULL) {
+      fieldsArray = prop_dictionary_get_keysym(dict, device);
+      if (fieldsArray == NULL)
+         goto error;
+
+      fieldsIter = prop_array_iterator(fieldsArray);
+      if (fieldsIter == NULL)
+         goto error;
+
+      isACAD = false;
+      isBattery = false;
+
+      /* only assume battery is not present if explicitly stated */
+      isPresent = 1;
+      isConnected = 0;
+      curCharge = 0;
+      maxCharge = 0;
+
+      while ((fields = prop_object_iterator_next(fieldsIter)) != NULL) {
+         props = prop_dictionary_get(fields, "device-properties");
+         if (props != NULL) {
+            class = prop_dictionary_get(props, "device-class");
+
+            /*
+             * After NetBSD 11's release NetBSD 9 will no longer be supported
+             * and these should be converted to prop_string_equals_string.
+             */
+
+            if (prop_string_equals_cstring(class, "ac-adapter")) {
+               isACAD = true;
+            } else if (prop_string_equals_cstring(class, "battery")) {
+               isBattery = true;
+            }
+            continue;
+         }
+
+         curValue = prop_dictionary_get(fields, "cur-value");
+         maxValue = prop_dictionary_get(fields, "max-value");
+         descField = prop_dictionary_get(fields, "description");
+
+         if (descField == NULL || curValue == NULL)
+            continue;
+
+         if (prop_string_equals_cstring(descField, "connected")) {
+            isConnected = prop_number_integer_value(curValue);
+         } else if (prop_string_equals_cstring(descField, "present")) {
+            isPresent = prop_number_integer_value(curValue);
+         } else if (prop_string_equals_cstring(descField, "charge")) {
+            if (maxValue == NULL)
+               continue;
+            curCharge = prop_number_integer_value(curValue);
+            maxCharge = prop_number_integer_value(maxValue);
+         }
+      }
+
+      if (isBattery && isPresent) {
+         totalCharge += curCharge;
+         totalCapacity += maxCharge;
+      }
+
+      if (isACAD) {
+         *isOnAC = isConnected ? AC_PRESENT : AC_ABSENT;
```

Comment:
What if there were multiple AC adapters present, but one disconnected?

Reply:
What do other OSes do in this case?

Reply:
Well, darwin leaks memory and takes the last one it finds … FreeBSD only seems to handle one battery, but depending on the return value of the syscall may treat everything with at least one connected source as on AC. Linux takes the ACPI reading for all batteries and terminates upon finding the first that's AC … Also the Linux code looks like it wants to receive some cleanups … PCP is not yet implemented … ;-) Neither is Solaris …

I think, the best behavior is to treat at least online AC connection as `AC_PRESENT` … Also don't leak memory while doing this. ;-)